translation: use gpt-4o-transcribe for speech-to-text

mamuqsit opened the pull request:
## Introduction
Audio responses face severe and unpredictable latency (P95 stretching into minutes). Investigation showed Whisper-based STT was the primary bottleneck - Whisper handles at most 3 requests per minute which is quite low for our utility.

Swapped Whisper with GPT-4o Transcribe (10,000 requests per minute) for English, Hindi, and Telugu languages.

## Changes
* Refactor Azure OpenAI Whisper → generic Azure OpenAI speech translator
* Add per-language STT/TTS configuration via app config
* MCP tool `asha_chat` now supports audio inputs
* Add integration tests to ensure audio message inputs do not yield IDKs
* Introduce `ASHABOT_FEATURE_FLAGS=stt_latency_mitigation` to enable gpt-4o-transcribe STT service for all en, hi, te users. By default, gpt-4o-transcribe service is only served to `User(test_user=True, user_language in {en, hi, te})`.

## Tests
Results indicate STT no longer hangs for 60 seconds on receiving the 4th chat message. The first block below uses Whisper backend, and the second block uses GPT-4o Transcribe backend. See below for Python code snippet to replicate the test.
```
$ python3 test_stt.py
Executing tool calls...
Executed tool calls in 77.55877089500427 seconds, got responses:
- What is the appropriate weight for a 3-year-old girl? -> 3 साल की लड़की का उचित वजन 13.9 किलोग्राम होता है।
- What is the appropriate weight for a 3-year-old girl? -> 3 साल की लड़की का उचित वजन 13.9 किलोग्राम होता है।
- What is the appropriate weight for a 3-year-old girl? -> 3 साल की लड़की का उचित वजन 13.9 किलोग्राम होता है।
- What is the appropriate weight for a 3-year-old girl? -> 3 साल की लड़की का उचित वजन 13.9 किलोग्राम होता है।
Executing tool calls...
Executed tool calls in 12.187675952911377 seconds, got responses:
- What is the normal weight for a three-year-old girl? -> तीन साल की लड़की का सामान्य वजन लगभग 13.9 किलोग्राम होता है।
- What is the normal weight for a three-year-old girl? -> तीन साल की लड़की का सामान्य वजन लगभग 13.9 किलोग्राम होता है।
- What is the normal weight for a three-year-old girl? -> तीन साल की लड़की का सामान्य वजन लगभग 13.9 किलोग्राम होता है।
- What is the normal weight for a three-year-old girl? -> तीन साल की लड़की का सामान्य वजन लगभग 13.9 किलोग्राम होता है।
```

<details>
<summary>test_stt.py</summary>

```py
import asyncio
import base64
import pickle
import requests
import time
from fastmcp import Client

URL = "http://127.0.0.1:8000"
PHONE_NUMBER_ID = "91xxxx"
USER_NAME = "Muqsit"

audio_data = requests.get("https://github.com/A4i-tech/byoeb/raw/3bc0c969c1f4af55b223489251aa0584d87f9161/byoeb-v1/byoeb/tests/integration/resources/hindi.ogg").content

async def test(test_user: bool):
    requests.delete(f"{URL}/delete_users", json=[PHONE_NUMBER_ID]).raise_for_status()
    response = requests.post(f"{URL}/register_users", json=[{
        "phone_number_id": PHONE_NUMBER_ID,
        "user_location": {"district": "Pytest District"},
        "user_type": "asha",
        "user_language": "hi",
        "user_name": USER_NAME,
        "test_user": test_user,
    }])
    response.raise_for_status()

    tool_name = "asha_chat"
    tool_params = {"message": {"data": base64.b64encode(audio_data), "mime_type": "audio/ogg"}}
    print("Executing tool calls...")
    async with Client(URL + "/mcp?phone_number=" + PHONE_NUMBER_ID) as client:
        # call tool 4 times concurrently
        tasks = [client.call_tool(tool_name, tool_params) for _ in range(4)]
        t = time.time()
        responses = await asyncio.gather(*tasks)
        t = time.time() - t

    print("Executed tool calls in", t, "seconds, got responses:")
    for response in responses:
        print("-", response.data.additional_info[0][1], "->", response.data.text)


asyncio.run(test(test_user=False))
asyncio.run(test(test_user=True))
```
</details>

## Follow-up tasks
- Create a deployment for `gpt-4o-transcribe` model on the Azure OpenAI service used in production
- Update environment variables in staging and production environment:
  - `AZURE_COGNITIVE_REGION=`
  - `AZURE_COGNITIVE_TEXT_TO_SPEECH_RESOURCE=`
  - `AZURE_COGNITIVE_TEXT_TO_TEXT_RESOURCE=`

mamuqsit committed:
refactor: rename async_azure_openai_whisper -> async_azure_openai_translator
This impl is generic for all OpenAI transcriber models, we can reuse it for GPT-4o Transcribe.

mamuqsit committed:
mcp: support audio input in asha_chat

mamuqsit committed:
dep-setup: suffix speech translators by their utility (_tts, _stt)

mamuqsit committed:
app-config: allow specifying per-language stt and tts translation services

mamuqsit committed:
app-config: use gpt-4o-transcribe for en, hi, te; whisper for mr
Goal was to use whisper for te as well, but whisper bails on processing te (unsupported)

a4i-architect reviewed (commented):
**PR Review Summary**

Overall, the code improvements reflect proactive efforts towards better type safety, cleaner media data handling, and updated dependencies. Key positive changes include the proper base64 decoding in `media_request.py` and the alignment of function signatures with expected data types in translation interfaces. However, several areas require attention:

---

### Strengths & Positive Aspects:
- **Type safety and clarity**: Transition to `field_validator` with `mode="before"` in `MediaData` is a solid move, reinforcing schema validation.
- **Code consistency**: Adjustments in method signatures to use `str` or `bytes` appropriately improve interface clarity.
- **Dependency injection update**: Replacing `speech_translator_whisper` with `speech_stt` in the message workflow enhances flexibility and aligns with new abstractions.
- **Configuration updates**: JSON config expansion integrates new speech services, improving configurability and future scalability.
- **Minimal duplicate imports**: Clarifications in imports, avoiding redundant code snippets.

---

### Areas for Improvement & Critical Feedback:

1. **Media Handling & Decoding**:
   - The media decoding logic in `media_request.py` is solid; however, consider adding validation that the incoming string is valid base64 before decoding to prevent unexpected errors.
   - Ensure the `bytes` attribute `data` is consistently validated across the entire flow, especially when passing images/audio.

2. **Interface & Signature Consistency**:
   - The change from `audio_file: str` to `audio_data: str` or `bytes` in translator classes is correct, but enforce strict validation and docstrings to prevent misuse.
   - For example, in `speech/azure/async_azure_openai_translator.py`, method signatures inconsistently mention `audio_data` or `audio_file`. Uniformity improves maintainability.

3. **Code Duplication & Overengineering**:
   - The large `if-elif` structure in `dependency_setup.py` for initializing speech services could be refactored into a cleaner Factory Pattern. This reduces cyclomatic complexity and improves maintainability.
   - Consider a registry-based approach rather than nested matches, which makes adding new services inherently simpler.

4. **Error Handling & Validation**:
   - The new base64 decoder should incorporate robust error handling.
   - Configuration checks such as `if env_config.env_azure_openai_speech_key:` often raise warnings but might benefit from explicit exception raising during startup if critical dependencies are missing.

5. **Code Readability & Complexity**:
   - Several functions, especially in `process.py` and `generate.py`, are becoming large and complex. Break down into smaller, testable components.
   - Around line 100+ in `process.py`, the logic handling media download, transcription, and query rewriting is intertwined. Use early returns and dedicated helper functions for clarity.

6. **Design & Pattern Usage**:
   - The code heavily relies on direct imports within functions, e.g., dependency injection functions, which could be abstracted further with proper class-based initialization.
   - The use of `TypeAdapter` for configs is good, but static type checking or validation at startup could spot mismatches early.

7. **Potential Semantic Bugs & Typos**:
   - Certain variable name inconsistencies: e.g., in `chat.py` the buffer `message` may sometimes be a `str` or a `MediaData`. Explicit type checking or overloading helps.
   - In media processing, ensure fallback behaviors for missing or corrupt media streams.

8. **Performance validation**:
   - The addition of timing logs in transcript annotation is good. Consider adding metrics or stats gathering for repeated transcriptions for performance tuning.

---

### Web Articles & Best Practices:
- [A Guide to Clean Code and Refactoring Complex Logic](https://refactoring.guru/refactoring) emphasizes that large `if-else` chains, such as in your service registry, should be replaced by polymorphism or factory patterns.
- [Type Safety in Python](https://realpython.com/static-types-in-python/) advocates for strict typing and validation, which your `pydantic` schemas are moving towards.
- [Design Pattern for Dependency Injection](https://python.plainenglish.io/dependency-injection-in-python-7d85a329fd1) to modularize configuration loading.

---

### Final Evaluation:
**Score: 75/100**

**Reasoning:**  
The code has progressed with meaningful improvements, notably in type safety, configuration management, and clearer method signatures. However, some patterns (large if-else chains, nested media handling, inconsistent validation) remain overly complex. For production-ready robustness, especially as new speech services are integrated, refactoring into more maintainable design patterns is recommended.

**Acceptable with recommendations to refactor for clarity and scalability.**

mamuqsit committed:
AzureOpenAIWhisperParamsEnum -> AzureOpenAISpeechParamsEnum

mamuqsit committed:
dep-setup: make speech_* require user

mamuqsit committed:
stt: gate gpt-4o-transcribe behind feature flag 'stt_latency_mitigation'

mamuqsit committed:
tests: add integration tests for en, hi audio inputs

a4i-architect reviewed (commented):
**PR Review Summary & Feedback**

Overall, the code modifications introduce several improvements for robustness, modularity, and clarity, especially around speech translation handling and media processing. Key positive points include:

1. **Consistent Base64 Decoding in Media Requests**: Adding a `@field_validator` for `data` in `MediaData` simplifies encoding handling and improves robustness.

2. **Refactoring of Speech Translator Interfaces**: Transitioning from specific `audio_file` and `bytes` parameters to `audio_data` strings/bytes and aligning method signatures (`speech_to_text`, `aspeech_to_text`) across subclasses enhances interface consistency and prepares for more flexible implementations.

3. **Improved Dependency Injection & Modular Initialization**: The dynamic setup for speech translators based on configuration (Azure, OpenAI Whisper, etc.) is much cleaner, leveraging type adapters, feature flags, and environment variables. This pattern aligns with best practices for scalable service registration ([ref](https://12factor.net/config)).

4. **Media Processing & Transcription Handling**: The addition of `annotate_audio_transcription()` in workflow handlers injects actual media download, conversion, transcription, and text assignment in a clear, centralized manner. This reduces duplicated code and simplifies asynchronous flows.

5. **Code Cleanup & JSON Configuration Corrections**: Enhancing the JSON config for speech services with explicit API versions and resource IDs increases flexibility, maintainability, and clarity.

---

### Areas for Improvement and Cautions:

#### 1. **Unnecessary or Overly Complicated Logic**
- **Signal of Cyclomatic Complexity Risks**: The service registration loop for speech translation creates multiple case branches, sometimes using nested `match` statements. This pattern can become hard to maintain. Consider utilizing a factory pattern or registration map instead of verbose `match ... case` blocks, reducing cyclomatic complexity ([Ref](https://refactoring.guru/design-patterns/factory)).  
- **Feature Flag Handling**: The gated logic for latency mitigation toggles-based on feature flags and environment, which is good. However, the conditional construction (`if env_config.env_azure_openai_speech_key:`) could be refactored into a dictionary-based registry for cleaner extensibility.

#### 2. **Typings & API Usages**
- The consistent use of `Optional[str]` for `source_language` in the interface is good, but some subclasses do not have explicit type annotations, especially in overridden methods. Make sure override signatures exactly match the base class to avoid runtime issues.
- In the media request validator, decoding base64 into bytes works fine, but consider adding explicit validation/error handling if string data is malformed or not base64. 

#### 3. **Redundant or Repetitive Code**
- The pattern for setting up speech services involves similar code blocks for checking API keys, environment variables, etc. Abstracting this setup into a dedicated factory or builder class would significantly reduce duplication and improve testability ([Ref](https://refactoring.guru/design-patterns/builder)).

#### 4. **Design & Maintainability**
- The large configuration-driven registration process for speech services is effective but can be made more robust by defining a registry object or data class for each service type, minimizing manual update errors.
- The `media_info` handling in `annotate_audio_transcription()` assumes the presence of `media_info.media_id`. Add error handling or validation in case of malformed/media missing data.

#### 5. **Testing & Coverage**
- The provided integration test for audio messages is a good baseline. You should extend tests to cover failure modes: invalid base64, corrupted files, missing media info, API failures, etc.

---

### Web Search Articles for Context
- **Implementing Factory Pattern for Service Registration**: [Ref](https://refactoring.guru/design-patterns/factory)
- **Best Practices for Config-Driven Service Setup**: [Ref](https://12factor.net/config)
- **Reducing Cyclomatic Complexity in Python**: [Ref](https://medium.com/swlh/5-ways-to-reduce-cyclomatic-complexity-in-python-9eb75c574d39)

---

### **Score: 78/100**

**Strengths**:
- Good structural refactorings and consistent method signatures.
- Clear separation of configuration-driven service registration.
- Enhanced media processing flow is clean and centralized.

**Improvements Needed**:
- Abstract repetitive service registration code into a factory.
- Simplify nested `match-case` blocks.
- Add validation/error handling around media decoding and external API calls.
- Better document assumptions, especially in media/media_info handling.

**Conclusion**: The current update significantly improves maintainability and integration flexibility but can still be optimized for simplicity and robustness. With minor refactoring and additional validation, it will meet higher standards of clean, scalable architecture.

---

**Recommendation**: Proceed with acceptance after addressing the iteration points on abstraction and error handling.

mamuqsit committed:
azure-speech: fix failing test

a4i-architect reviewed (commented):
**PR Review Summary:**

**Overall Impression:**  
The codebase has seen thoughtful improvements, especially around handling media data, async translation classes, and environment-based configuration. However, there are several areas that could benefit from simplification, stricter typing, and elimination of code duplication. The complexity, particularly in the speech translation setup and media processing workflows, can be reduced for better maintainability.

---

### File-by-File Analysis:

#### 1. **media_request.py**
- Good inclusion of base64 decoding via `field_validator`.  
- No major issues, but consider renaming `decode_base64_if_needed` to clarify the intention (e.g., `decode_base64_bytes`).  
- Ensure that downstream consumers are aware that `data` is always bytes now.

#### 2. **speech/base.py**
- The switch from `audio_file: str` to `audio_data: str` is correct for clarity.  
- Slight inconsistency: method signatures for `speech_to_text` vs `aspeech_to_text`—avoid confusion by standardizing naming conventions or documenting method behavior thoroughly.  
- Abstract methods are well-typed; however, considering return types and exceptions explicitly could improve clarity.

#### 3. **async_azure_openai_translator.py**
- The restructuring to avoid duplicated code in initializing `AsyncAzureOpenAI` is positive.  
- The main concern is the verbosity and repeated pattern when setting up clients. Could benefit from a factory/helper function, reducing cyclomatic complexity.  
- Asynchronous workflow is correctly employed.  
- Typings are now clearer, and environment variable handling is consistent.

#### 4. **async_azure_speech_translator.py**
- Logical voice mappings are well-structured.  
- Method signatures are consistent.  
- Minor: The current `speech_to_text` stub could throw `NotImplementedError`—fine if intended.  
- Good test coverage indicated.  

#### 5. **test_translators_azure_speech.py**
- The tests are comprehensive, covering missing configuration, errors, and success paths.  
- Use of `AsyncMock` is appropriate.  
- The naming of variables and test structure is clean.  
- Minor: Repeated code snippets for client setup could be refactored into fixtures.

#### 6. **chat.py (build method & asha_chat)**
- Type annotations improved, especially allowing `MediaData` in `message`.  
- The `asha_chat` logic correctly integrates media handling.  
- The pattern of handling media in workflows is robust but somewhat verbose; consider encapsulating repetitive media fetching logic.  
- The use of `message.message_context.message_type == MessageTypes.REGULAR_AUDIO.value` for media detection is sound.  
- The media handling and transcription logic could be modularized further.

#### 7. **respond_with_consensus.py**
- The update to import `MediaData` and the refactoring of `create_user_message` enhance clarity.  
- The addition of `annotate_audio_transcription()` and its integration is well-done.  
- The code is straightforward; however, `audio_message` fetch and conversion could be moved out for testability.

#### 8. **config.json & dependency_setup.py**
- Configuration JSON upgrades to include multiple speech services, which improves flexibility.
- The feature flag addition (`STT_LATENCY_MITIGATION`) is a strategic enhancement.  
- The environment variable handling for keys is improved, but the code could be further DRYed by abstracting environment checks into a helper.

#### 9. **constants/feature_enums.py**
- Small but meaningful addition of a feature flag related to latency mitigation—good practice.

#### 10. **related handler and processing files**
- The use of dependency injections and context passing is consistent.
- Media handling in `process.py` is expanded. Critical: The `annotate_audio_transcription()` now encapsulates media fetch and conversion, reducing clutter.
- Exception/error handling should further be enforced to prevent silent failures.

#### 11. **tests/integration/test_audio_message.py**
- Good test coverage with environment setup.
- Asynchronous test functions are well-structured.
- Suggest improvements in dynamic user creation—add teardown or cleanup to prevent pervasive test artifacts.

---

### **General Recommendations & Observations:**

- **Code Duplication:**  
  Numerous patterns for initializing Azure services (speech/translation) are duplicated. Abstract into helper factories to reduce cyclomatic complexity and improve maintainability.  
  *Web resource:* [Refactoring To Factory Pattern](https://refactoring.guru/design-patterns/factory-method)

- **Type Safety & Consistency:**  
  Leverage more precise typing, e.g., for return types and APIs, to prevent runtime errors. Use `Optional` and `Union` explicitly.  

- **Media Handling & Workflow Complexity:**  
  The media download, convert, and transcription workflow is somewhat scattered. Encapsulating into dedicated services reduces complexity.

- **Configuration Management:**  
  JSON configs efficiently handle multiple service endpoints, but code will benefit from a centralized loader with validation steps.

- **Async Management & Thread Safety:**  
  The usage of `asyncio.Lock()` indicates concurrency considerations. Ensure proper scope and initialization—perhaps move lock creation into a dedicated class or singleton.

- **Logging & Error Handling:**  
  Consider adding more try-except blocks and logging exceptional states. This is especially critical for media download and external API calls.

- **Documentation & Naming:**  
  Clarify method names which sometimes differ subtly (`speech_to_text` vs `aspeech_to_text`). Consistent naming standards improve code understandability.

---

### **Score & Final Remarks:**

**Score:** 75/100

While the code is moving in the right direction, especially with better async handling and configuration management, it still exhibits complex conditional logic, duplicated setup patterns, and scope for more modular design. A dedicated effort around abstraction, cleaner error management, and standardized naming would push the score higher.

---

### **References:**

- [Refactoring to Factory Pattern](https://refactoring.guru/design-patterns/factory-method)
- [Effective Type Annotations](https://mypy.readthedocs.io/en/stable/cheat_sheet_py3.html)
- [Async Programming Patterns in Python](https://realpython.com/async-io-python/)

**In conclusion:**  
The improvements are commendable, but the architecture requires more modularity and abstraction to be production-ready. Focus on reducing code duplication, enhancing type safety, and modularizing complex workflows.

mamuqsit committed:
tests/integrations: add sample ogg files for audio messages

a4i-architect reviewed (commented):
**PR Review Summary:**

Overall, the codebase has undergone extensive refactoring focused on enhancing speech translation abstractions, JSON handling, and configuration management. Major improvements include replacing custom base64 decoding with pydantic’s `field_validator`, cleaning up redundant or inconsistent variable naming, and standardizing interface contracts across speech translation classes. 

**Positive observations:**
- The introduction of `field_validator` for decoding base64 is well-aligned with modern pydantic best practices.
- Transition toward a more dynamic, config-based approach for speech service initialization (e.g., using dictionaries and factory-like patterns) enhances scalability.
- Explicit typing, especially with `Optional` and union types, improves code clarity.
- Environment variable handling is more robust, accommodating fallback mechanisms.
- The testing suite for asynchronous speech translation behaviors appears comprehensive, including error cases and missing credentials.

**Areas for improvement:**

1. **Code Duplication & Overly Complex Logic:**
   - The configuration-based service instantiation in `dependency_setup.py` is quite verbose. Using factory patterns or a registry can reduce boilerplate.
   - The `speech_selector` lambda pattern is concise but could benefit from explicit class-based strategy patterns to improve testability and clarity.
   - There are nested `match ... case` statements that could become unwieldy with more service types; consider refactoring into dedicated factory functions.
   
2. **Type & Implementation Accuracy:**
   - In `speech/base.py` and related classes, method signatures switch between `str` and `bytes` without explicit conversion needs documented. Clarify if the `audio_data` is base64-encoded strings or raw bytes, and ensure consistent handling.
   - In the AWS/Azure classes, some variables are named inconsistently (`azure_endpoint`, `endpoint`, `api_key`, `token_provider`), leading to potential bugs or confusion.
   
3. **Design & Complexity:**
   - The `async def annotate_audio_transcription()` method in message processing is tightly coupled to media download and conversion but can be abstracted further or moved into dedicated service classes.
   - The `handle_process_message_workflow()` combines multiple responsibilities: downloading media, transcription, translation, rewritting, which increases cyclomatic complexity. Consider splitting into smaller handlers.
   
4. **Configuration & Environment Handling:**
   - There is repetition in checking whether API keys are set, which can be unified into utility functions.
   - The per-language service map (e.g., `_speech_stt`) coupled with dynamic registration is clever but risks runtime errors if a language is missing; add validation post-initialization.
   
5. **Testing & Coverage:**
   - The test for audio message is a good functional coverage, but more boundary cases (missing media, failed download, unsupported MIME types) should be added.
   - Mocks for external media download and translation services are appropriately used, but consider parameterizing tests further for edge cases.

**Score: 75/100**

The changes are well-intentioned with sound architectural intent, but the overall complexity and some inconsistent naming/handling reduce maintainability and robustness. For a premium codebase, simplifying or abstracting repetitive patterns, improving error handling, and delineating responsibilities more cleanly would elevate the quality further.

**Recommended articles for detail:**
- [Best practices for handling external API configurations](https://realpython.com/handling-secret-keys-in-python/)
- [Design patterns for factory initialization: Strategy pattern](https://refactoring.guru/design-patterns/strategy)
- [Using pydantic validators for data transformation](https://pydantic.dev/usage/validation_decorator/)

**In summary:**
- Solid groundwork with modern paradigms.
- Needs focus on reducing complexity, standardizing patterns, and improving test robustness.
- Final score: **75/100** (meets minimum acceptance threshold).

Soumabha commented on byoeb-v1/byoeb/byoeb/chat_app/configuration/config.py:
```diff
@@ -52,7 +52,8 @@
 env_azure_speech_key= os.getenv("AZURE_SPEECH_KEY")
-env_azure_openai_whisper_key= os.getenv("AZURE_OPENAI_WHISPER_KEY")
+env_azure_openai_speech_key = os.getenv("AZURE_OPENAI_SPEECH_KEY") or os.getenv("AZURE_OPENAI_WHISPER_KEY")
+env_azure_openai_speech_endpoint = os.getenv("AZURE_OPENAI_SPEECH_ENDPOINT") or "https://khushi-baby-asha-oai.openai.azure.com/"
```
No defaults leading to production environment connection, throw error if not configured like other setup.

mamuqsit replied:
Addressed by https://github.com/A4i-tech/byoeb/pull/99/commits/25728cec533fdc6bb4d633250c1672a71ec1c492

Soumabha commented on byoeb-v1/byoeb/byoeb/chat_app/configuration/dependency_setup.py:
```diff
@@ -160,65 +166,56 @@ async def get_leaderboard_service() -> LeaderboardService:
-    azure_endpoint=app_config["translators"]["speech"]["azure_oai"]["endpoint"],
-    api_version=app_config["translators"]["speech"]["azure_oai"]["api_version"]
-)
+from byoeb_integrations.translators.speech.azure.async_azure_openai_translator import AsyncAzureOpenAISpeechTranslator
```
We should have the dependency setup inject only a speech adapter file. The speech adapter file can have the specific business logic for voice dictionary mapping as well as the multi adapter configuration.

mamuqsit replied:
Addressd in https://github.com/A4i-tech/byoeb/pull/99/commits/c37df03652f0e73564d5fbd59495ddcea82620c8 - `TranslatorAdapter` delegates tasks to children:
```
class TranslatorAdapter(BaseSpeechTranslator):
    stt: <Lang, BaseSpeechTranslator>
    tts: <Lang, BaseSpeechTranslator>

    def stt(lang, text) -> self.stt[lang].stt(text)
    def tts(lang, text) -> self.tts[lang].tts(text)
```

Soumabha replied:
Resolving it for now. From the future, best to make the "key" a combination of all the variables; in this case it would be langCode + gatedflag

mamuqsit replied:
Perhaps it would be cleaner to let the config itself govern reusability logic. E.g., define providers once and reference them by ID.
```json
{
    "providers": {
        "voice-model-v1": {
            "type": "azure-cognitive-services",
            "api_key": "..."
        },
        "transcribe-model-v1": {
            "type": "azure-openai",
            "api_key": "...",
            "model": "gpt-4o-transcribe"
        }
    },
    "capabilities": {
        "text_to_speech": {
            "provider": "voice-model-v1",
            "supported_languages": ["en", "hi", "mr", "te"]
        },
        "speech_to_text": {
            "provider": "transcribe-model-v1",
            "supported_languages": ["en", "hi", "te"]
        }
    }
}
```

Soumabha commented on byoeb-v1/byoeb/byoeb/chat_app/configuration/dependency_setup.py:
```diff
@@ -160,65 +166,56 @@ async def get_leaderboard_service() -> LeaderboardService:
+            attributes["api_key"] = env_config.env_azure_openai_speech_key
+            attributes["endpoint"] = env_config.env_azure_openai_speech_endpoint
+            speech_group = _speech_stt_gated if gated else _speech_stt
+            speech_group.update((lang, AsyncAzureOpenAISpeechTranslator(**attributes)) for lang in languages)
```
We seem to be creating an instance per translator.
To minimize the app memory footprint, we should have a single translator itself per type and use that as per user.language.

mamuqsit replied:
Addressed in https://github.com/A4i-tech/byoeb/pull/99/commits/6220ea8ef20dddf82fce3c11c36759183147e2f3 - we can hash(config) of these models and reuse models having same config.

Soumabha commented on byoeb-v1/byoeb/byoeb/chat_app/app_config.json:
```diff
@@ -80,17 +80,37 @@
+                    "region": "swedencentral"
+                }
+            },
+            {
```
Considering that we have multiple JSONs for the same languages, how does it match to use the appropriate one? 

Please check if we can update the api_versions of any of these models.

mamuqsit replied:
Addressed in https://github.com/A4i-tech/byoeb/pull/99/commits/64ac309411ca6bb73756e00bcac1c349893ead77, bumped api version 2024-06-01 -> 2024-10-21. It appears this isn't just the latest but the last api version for AzureOpenAI.

Azure suggests moving to Azure Foundry - with the Foundry models we get to directly utilize OpenAI() instead of AzureOpenAI().

mamuqsit replied:
> Considering that we have multiple JSONs for the same languages, how does it match to use the appropriate one?

Speech config implicitly overrides (order-dependent) - last entry in the "speech" array is the selected model for that set of languages:
```
models = {}
for entry in config["service"]:
    service = AzureOpenAITranslator(...)
    for lang in entry["languages"]:
        models[lang] = service
```

However, this configuration in app_config.json doesn't cause any overrides - every language has 1 model (if you ignore the `gated: true` entries).

Gated model are served to test users only (unless feature flag is set) and have priority over non-gated models.

Soumabha commented on byoeb-v1/byoeb/byoeb/chat_app/configuration/dependency_setup.py:
```diff
@@ -160,65 +166,56 @@ async def get_leaderboard_service() -> LeaderboardService:
+    return gated[lang] if user.test_user and lang in gated else default[lang]
+
+speech_stt: Callable[[User], BaseSpeechTranslator] = lambda u: _speech_selector(u, _speech_stt, _speech_stt_gated)
+speech_tts: Callable[[User], BaseSpeechTranslator] = lambda u: _speech_selector(u, _speech_tts, _speech_tts_gated)
```
I was considering just extending the existing setup to take test_user flag in addition to language code and auto decide the model based on that.

mamuqsit replied:
Addressed in https://github.com/A4i-tech/byoeb/pull/99/commits/e3616713e59b48cb3a791535185f4fc73fa0785e, superseded by https://github.com/A4i-tech/byoeb/pull/99/commits/c37df03652f0e73564d5fbd59495ddcea82620c8

mamuqsit committed:
Merge remote-tracking branch 'origin/a4i/staging' into muqsit/improve-audio-latency

mamuqsit committed:
AZURE_OPENAI_SPEECH_ENDPOINT: bail when undefined

mamuqsit committed:
speech: reuse models having identical configuration

mamuqsit committed:
speech_* selectors: accept lang, test_user instead of bulky User object for matching

mamuqsit committed:
app-config: update speech model api versions 2024-06-01 -> 2024-10-21

mamuqsit committed:
refactor: introduce a singular translator adapter for stt and tts

a4i-architect reviewed (commented):
**PR Review Summary**

Overall, these code updates involve enhancements to media request handling, speech translation abstraction, and test coverage. The core intention seems to be consolidating speech translation services, improving media processing, and refining integrations. Here's a detailed critique:

### Strengths & Positive Changes:
- **Refactoring & Abstraction of Speech Services:** The `TranslatorAdapter` class in `translator.py` centralizes speech translation services, supporting multiple providers and environments. This pattern improves maintainability and scalability.
- **Base64 Media Handling:** The `MediaData` model now decodes base64 strings before validation, simplifying downstream media processing.
- **Consistent Typing and Optionality:** Typing annotations are clarified (`str` vs `bytes`) in speech translator interfaces, reducing ambiguity.
- **Enhanced Test Coverage:** The test `test_audio_message.py` seems robust, involving real binary resources, and tests different language scenarios.
- **Configuration & Environment Variables:** External dependencies and secrets are cleaned up, and environment fallbacks are handled more gracefully.
- **Additional enums and flags:** Feature flags for latency mitigation and configs for speech source types provide flexible control.

---

### Points for Improvement & Critical Feedback:

1. **Code Duplication & Simplification:**
   - The `TranslatorAdapter` class maintains multiple dictionaries for gating/test user modes. This could be simplified using a unified configuration with parameters. Avoid duplicated logic in `__select_group` and `__select`.
   - The `__instantiate` method has verbose pattern matching. Consider using a factory pattern to decouple provider-specific logic further. For example, a dedicated factory class or registry.

2. **Design & Complexity:**
   - The `__instantiate` method has multiple pattern matches with nested conditionals. It can be simplified using a registration-based factory or Strategy pattern, reducing cyclomatic complexity.
   - The code switches between runtime environment detection (`env_config`) and hardcoded behaviors inconsistently—standardize environment handling for clarity.
   - Some parts, like speech translation and media processing, could be further abstracted into dedicated classes/services, notably for better testing and separation of concerns.

3. **Type Safety & Validation:**
   - In `MediaData`, `data` is now `bytes`, which is good. But the validator that decodes base64 should include error handling for invalid base64 data to avoid silent failures.
   - In methods like `aspeech_to_text`, parameter annotations differ between `str` and `bytes`. Consistency and precise typing are recommended.

4. **Logic & Clarity:**
   - The logic for gating services (`gated`) based on feature flags and test user mode is somewhat convoluted. Consider explicit flags or a dedicated gating component.
   - The `annotate_audio_transcription` method relies on downloading media from a media ID. This introduces potential side effects; consider making media fetching more explicit and testable via dependency injection.

5. **Test Coverage:**
   - Tests correctly cover different languages and environment configurations, but testing failure paths (e.g., invalid base64, missing environment variables) is lacking.
   - The integration test for audio is good but may require more edge cases for robustness.

6. **Configuration & Environment Handling:**
   - The JSON config modifications are fairly straightforward but should be validated to prevent runtime errors due to undefined keys.
   - Ensure that environment variables related to endpoints and keys are consistently handled across all components.

7. **Miscellaneous:**
   - Some commented-out code snippets and TODOs (e.g., factory in `translator.py`) indicate ongoing refactoring. Removing dead code or completing the factory pattern would be beneficial.

---

### Web Articles for Reference:
- [Factory Pattern in Python](https://refactoring.guru/design-patterns/factory-method/python/example)
- [Reducing Cyclomatic Complexity](https://www.lucashalbert.com/2015/09/13/reduce-cyclomatic-complexity/)
- [Best Practices for Abstracting External Services](https://krasimirov.com/blog/clean-code/decoupling)

---

### Final Score: **75/100**

**Comments:** The code shows deliberate improvements toward modularity and configurability, especially with the `TranslatorAdapter`. However, it still retains overly complex pattern matching and duplicated logic, particularly in the factory instantiation methods. With refactoring towards a more pattern-driven factory approach, clearer separation of concerns, and more comprehensive testing, the code could reach a higher professionalism level.

**Recommendation:** Approve with suggestions for refactoring the factory logic, simplifying gating mechanisms, and improving test coverage, especially for error conditions.

mamuqsit committed:
fix failing test case
We shouldn't be using dependency_setup.py in our unit tests, but this is sufficient for time being

a4i-architect reviewed (commented):
**PR Review Summary**

**Overall Impression:**  
The codebase exhibits significant improvements in recent refactors, particularly around the abstraction of speech translation services and the enhancement of media handling. The modularization, especially in `TranslatorAdapter`, improves maintainability and scalability. However, some choices introduce potential pitfalls, and certain areas remain overly complex or inconsistent, impacting overall code clarity and robustness.

---

### Strengths:
- **Better abstractions:** The introduction of `TranslatorAdapter` streamlines speech API integrations, reducing duplicated code.
- **Enhanced media decoding:** Handling base64 decoding and media transformations (ogg to wav) are properly abstracted with `field_validator` and utility functions.
- **Config-driven approach:** JSON-based service configurations with flexible environments support easier deployments and testing.
- **Test setup improvements:** Environment variable mocking and test fixture setup are more consistent.

---

### Key Issues & Opportunities:

1. **Redundant & Ill-typed Code:**
   - In `media_request.py`, `decode_base64_if_needed` is a concise and correct pattern, but verifying the decoder’s robustness against malformed data is recommended.
   - In `utils.py`, `hash_dict` is solid, but be cautious of using `json.dumps()` — consider canonical forms like `json.dumps(d, sort_keys=True, ... )` to prevent subtle differences affecting caching.

2. **Design & Complexity:**
   - The `TranslatorAdapter` implements lots of domain logic (e.g., selecting gated vs. nongated services, multiple conditions). While flexible, it can be simplified further by implementing dedicated factory classes.
   - Cyclomatic complexity remains high in some methods, especially `__instantiate` and `__select_group`. Using strategy patterns or a lookup dict with callable factories increases readability.

3. **Code Duplication & Over-Handling:**
   - `speech_to_text` and `atext_to_speech` methods repeat a pattern of service selection. Extracting a simple dispatcher could reduce boilerplate.
   - The environment variable logic in `dependency_setup.py` for keys and endpoints is verbose. A centralized factory with validation would improve readability.

4. **Potential Misuse & Type Safety:**
   - The `aspeech_to_text` method expects `audio_data: bytes`, but in the Azure translator, you've changed the signature to accept `str`. Consistency should be enforced throughout.
   - The `MediaData` decoder in `media_request.py` assumes well-formed input; consider adding exception handling.

5. **Unit Tests & Integration:**
   - The test code, especially `test_audio_message.py`, correctly covers media-handling scenarios but should include more edge cases (e.g., malformed base64, unsupported media types).
   - Mocking external services (Azure SDKs, OpenAI) is partial; consider using dedicated mocks/stubs for better isolation.

6. **Configuration Files & Secrets:**
   - JSON configs are well-structured, but sensitive data injection (like API keys) via environment variables can be further validated at startup to prevent runtime errors.

7. **Additional Observations:**
   - The `feature_enums.py` shows a new feature flag `STT_LATENCY_MITIGATION`. Ensure this flag is handled gracefully across all speech modules.
   - Comments in `dependency_setup.py` indicate 'TODO: factory implementation'; planning to fully implement factories will enhance code clarity.

---

### Recommendations:
- **Refactor complex methods** for clarity, e.g., extract service factories, use lookup dictionaries for speech types/services.
- **Add explicit error handling** in media decoding and audio conversions.
- **Enforce consistent function signatures** across `async`/`sync` methods, especially for `audio_data` parameters.
- **Introduce unit tests** around provider selection, especially edge cases with missing or invalid environment variables.
- **Document architecture**: Brief diagrams or comments about the service selection pattern would help onboard new developers.
- **Consider Web Search Articles**:
  - [Design Patterns for Service Factories](https://refactoring.guru/design-patterns/factory)
  - [Managing Runtime Configuration & Secrets](https://12factor.net/config)
  - [Best Practices for Media Handling in Python](https://pydub.com/)

---

### Final Score: **78/100**  
The recent improvements are notable and mostly aligned with scalable, modular design, but the code still suffers from over-complication, inconsistent typing, and potential runtime pitfalls that prevent a higher score.

---

Soumabha commented on byoeb-v1/byoeb/byoeb/chat_app/app_config.json:
```diff
@@ -80,17 +80,37 @@
+                "service": "azure_cognitive",
+                "languages": ["en", "hi", "mr", "te"],
+                "attributes": {
+                    "resource_id": "/subscriptions/fa06c841-cd7a-4855-b39b-70621812fff3/resourceGroups/khushi_baby_asha/providers/Microsoft.CognitiveServices/accounts/khushi-baby-asha-speech-to-text",
```
Considering that we would ideally like to have separate resources possible for different environments - move to environment variables

mamuqsit replied:
Addressed in https://github.com/A4i-tech/byoeb/pull/99/commits/3751dcef086a38a9b3d7cdf5b2653b49ad985fae

mamuqsit committed:
dep-setup: read azure cognitive resource ids from env

a4i-architect reviewed (commented):
This comprehensive set of code modifications spans multiple modules, mainly enhancing media request handling, speech translation, and configuration. The core themes are better type safety, cleaner abstraction, improved dependency injection, and environment management.

**Positive Aspects & Strengths:**
- Transition from string to bytes for media data (`media_request.py`) with a decoder validator improves robustness.
- Dependency injection pattern in speech translation classes enhances testability.
- Environment variables handling in `dependency_setup.py` centralizes configuration, reducing hardcoded values.
- Adding `hash_dict()` for config versioning aligns with best practices for cache management.
- Tests revamped to better mock and isolate environment dependencies, with proper environment variable setups, are a big plus for test reliability.

**Key Weaknesses & Areas for Improvement:**

1. **Complex Conditional Logic & Over-Engineering:**
   - The `TranslatorAdapter` class (in `translator.py`) is quite complex, with nested `match` clauses and caching logic which makes it hard to follow. It could be simplified:
     - Use explicit factory functions for each service configuration.
     - Reduce the reliance on `hash_dict()` and instead install a cleaner registry.
     - The selection logic for gated vs default is somewhat convoluted, which could be abstracted better or clarified with explicit mappings.
2. **Type Safety & Consistency:**
   - Many functions, like `speech_to_text` and `atext_to_speech`, are loosely typed. Adding stricter type hints would improve static analysis.
   - Some methods always accept parameters like `test_user`, but it's optional or defaults, which should be clearly specified or eliminated if unnecessary.
3. **Simplify and DRY:**
   - The code for environment variable validation is repetitive. Consider centralizing validation routines and raising exceptions early.
   - Duplicate structures for Azure services can be unified through configuration-driven classes or a registry pattern.
4. **Error Handling & Failures:**
   - While many environment validations are present, error handling around failed external service calls is missing.
   - The code raising `RuntimeError` directly could be replaced with more semantic exceptions.
5. **High Cyclomatic Complexity & Readability:**
   - The `__instantiate()` method's match-case is lengthy and twisted. Instead, segregate each service configuration into dedicated factory functions.
   - Use of pattern matching, while elegant, could be replaced with straightforward if-else for clarity.
6. **Redundant Code or Over-Verbose Comments:**
   - Comments such as `# TODO factory implementation` hint at incomplete refactoring.
   - Some commented-out code from previous translation (like whisper) indicates mixed states; cleanup is recommended.
7. **Test Stability & Environment:**
   - The `conftest.py` sets environment variables directly, which is fine, but it might be more robust to mock environment at build time.
   - Test files seem comprehensive but could include more edge cases, e.g., invalid audio, missing env configs.

**Design Insights & Patterns:**
- The use of dependency injection in speech translators is a sound pattern, allowing flexible swapping of implementations.
- The new factory pattern in `TranslatorAdapter` is promising but over-complicated; employing a registration-based factory would be more maintainable.
- The abstraction level for different cloud services could be standardized, possibly creating an interface or base class for all speech services.

**Overall Score: 73/100**
- Meets many best practices but needs refinements in code clarity, abstraction, and error management before it can be considered truly robust and maintainable.  
- The code is somewhat overly complex for the given use-case, which could hinder future scalability.

**Related articles for best practices:**
- [Factory Pattern in Python](https://refactoring.guru/design-patterns/factory-method/python/example)
- [Config Management & Environment Variables](https://12factor.net/config)
- [Reducing Cyclomatic Complexity](https://refactoring.guru/common-principles/complexity)  

**Summary:**  
The code modifications are generally beneficial, particularly in environment configuration, media processing, and speech translation abstractions. However, the implementation of the `TranslatorAdapter` pattern is over-engineered. Simplifying conditional branches, centralizing factory logic, and ensuring consistent type safety will significantly improve maintainability. Also, enhancing error handling around external integrations will bolster robustness. With these improvements, the codebase can achieve higher clarity and reliability, aligning perfectly with software engineering best practices.
